projects/reservoir/ks.py

```
@@ -116,51 +119,85 @@ def _generate_ic(input_dim, seed=0):
     """
 
     N: int
     domain_size: int
     timestep: float
     spatial_downsampling_factor: int = 1
-    time_downsampling_factor: int = 1
+    error_eps: float = 0.0
     subdomain_output_size: Optional[int] = None
     subdomain_overlap: Optional[int] = None
     subdomain_axis: int = 1
 
-    def generate(self, n_steps: int, seed: int = 0):
+    def generate_from_seed(self, n_steps: int, seed: int = 0):
         """
-            n_steps: Number of timesteps to output. The interval between
-                timesteps output is timestep * time_downsampling_factor
+            n_steps: Number of timesteps to output
             seed: Random seed for initial condition
 
         Returns:
             Time series array with dims (time, x)
         """
-        ks_time_series = generate_ks_time_series(
-            input_size=self.N * self.spatial_downsampling_factor,
-            n_steps=n_steps * self.time_downsampling_factor,
-            seed=seed,
-            domain_size=self.domain_size,
-            timestep=self.timestep,
-        )
-
-        ks_time_series = ks_time_series[:: self.time_downsampling_factor, :]
-        ks_time_series = ks_time_series[:, :: self.spatial_downsampling_factor]
+        ic = _generate_ic(self.N * self.spatial_downsampling_factor, seed=seed)
+        ks_time_series = self.predict(n_steps=n_steps, initial_condition=ic)
 
         return ks_time_series
+
+    def predict(self, n_steps: int, initial_condition: np.ndarray):
+        if initial_condition.size != self.spatial_downsampling_factor * self.N:
+            # upsample initial condition array by spatial_downsampling_factor
+            ic_x = range(initial_condition.size)
+            interp_ic = interp1d(ic_x, initial_condition)
+            initial_condition = interp_ic(
+                np.linspace(
+                    ic_x[0], ic_x[-1], self.N * self.spatial_downsampling_factor
+                )
+            )
```

I might handle this interpolation explicitly within `create_imperfect_prediction_train_data` and maybe instead raise an error here if the initial condition size does not match the appropriate size based on the KS parameters.  Handling interpolation automatically is convenient, but I think it could also mask potentially unintended bugs.